prov/usnic/src/usdf_dgram.c

```diff
@@ -124,204 +282,146 @@ usdf_dgram_recvv(struct fid_ep *fep, const struct iovec *iov, void **desc,
+		void *desc, fi_addr_t dest_addr, void *context)
 {
-	struct usdf_ep *ep;
 	struct usdf_dest *dest;
+	struct usdf_ep *ep;
 
 	ep = ep_ftou(fep);
-
 	dest = (struct usdf_dest *)(uintptr_t) dest_addr;
-	return _usdf_dgram_send(ep, dest, buf, len, context);
-}
 
-ssize_t
-usdf_dgram_senddata(struct fid_ep *fep, const void *buf, size_t len,
-			void *desc, uint64_t data, fi_addr_t dest_addr,
-			void *context)
-{
-	USDF_TRACE_SYS(EP_DATA, "\n"); /* XXX delete once implemented */
-	return -FI_ENOSYS;
+	if (len + sizeof(struct usd_udp_hdr) <= USD_SEND_MAX_COPY) {
+		return usd_post_send_one_copy(ep->e.dg.ep_qp, &dest->ds_dest,
+						buf, len, ep->ep_tx_completion,
+						context);
+	} else if (ep->e.dg.tx_op_flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA,
+				"given inject length (%zu) exceeds max inject length (%d)\n",
+				len + sizeof(struct usd_udp_hdr),
+				USD_SEND_MAX_COPY);
+
+		return -FI_ENOSPC;
+	}
+
+	return usd_post_send_one(ep->e.dg.ep_qp, &dest->ds_dest, buf, len,
+				ep->ep_tx_completion, context);
 }
 
-static ssize_t
-_usdf_dgram_send_iov_copy(struct usdf_ep *ep, struct usd_dest *dest,
-		const struct iovec *iov, size_t count, void *context)
+ssize_t usdf_dgram_sendv(struct fid_ep *fep, const struct iovec *iov,
+		void **desc, size_t count, fi_addr_t dest_addr, void *context)
 {
-	struct usd_wq *wq;
- 	struct usd_qp_impl *qp;
-	struct usd_udp_hdr *hdr;
-	uint32_t last_post;
-	struct usd_wq_post_info *info;
-	uint8_t *copybuf;
+	struct usd_dest *dest;
+	struct usdf_ep *ep;
 	size_t len;
-	unsigned i;
 
-	qp = to_qpi(ep->e.dg.ep_qp);
-	wq = &qp->uq_wq;
-	copybuf = wq->uwq_copybuf +
-			wq->uwq_post_index * USD_SEND_MAX_COPY;
+	ep = ep_ftou(fep);
+	len = sizeof(struct usd_udp_hdr);
+	dest = (struct usd_dest *)(uintptr_t) dest_addr;
 
-	hdr = (struct usd_udp_hdr *)copybuf;
-	memcpy(hdr, &dest->ds_dest.ds_udp.u_hdr, sizeof(*hdr));
-	hdr->uh_udp.source =
-		qp->uq_attrs.uqa_local_addr.ul_addr.ul_udp.u_addr.sin_port;
+	len += _usdf_iov_len(iov, count);
 
-	len = sizeof(*hdr);
-	for (i = 0; i < count; i++) {
-		memcpy(copybuf + len, iov[i].iov_base, iov[i].iov_len);
-		len += iov[i].iov_len;
+	if (len <= USD_SEND_MAX_COPY) {
+		return _usdf_dgram_send_iov_copy(ep, dest, iov, count, context,
+						ep->ep_tx_completion);
+	} else if (ep->e.dg.tx_op_flags & FI_INJECT) {
+		USDF_DBG_SYS(EP_DATA,
+				"given inject length (%zu) exceeds max inject length (%d)\n",
+				len, USD_SEND_MAX_COPY);
+		return -FI_ENOSPC;
 	}
 
-	/* adjust lengths */
-	hdr->uh_ip.tot_len = htons(len - sizeof(struct ether_header));
-	hdr->uh_udp.len = htons(len - sizeof(struct ether_header) -
-			sizeof(struct iphdr));
-
-	last_post = _usd_post_send_one(wq, hdr, len, 1);
-
-	info = &wq->uwq_post_info[last_post];
-	info->wp_context = context;
-	info->wp_len = len;
+	/* _usdf_dgram_send_iov allocates buffer of size USDF_DGRAM_MAX_SGE
+	 * but first slot is for header.
+	 */
+	if (count > (USDF_DGRAM_MAX_SGE - 1)) {
```

Comment:
From bturrubiates/libfabric@5fde956c commit diff:

>Also, should this comparison be against USDF_DGRAM_DFLT_SGE (or some value derived from it) or txattr->iov_limit?

>In general though, fixing that needn't be a blocker for accepting this change. This is still a big improvement over where we were without this PR.

How about I move that check into `_usdf_dgram_send_iov` and change it to use `txattr->iov_limit` then change the way `iov_limit` is being set in a separate PR. Currently, it's being set to `USDF_DGRAM_DFLT_SGE` if no hint is provided.

I think it should be `USDF_DGRAM_DFLT_SGE` for prefix mode and `USDF_DGRAM_DFLT_SGE - 1` for non-prefix mode. Thoughts?